cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// Cluster is an API object representing a cluster's control-plane
+// configuration and status.
+type Cluster struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Spec   ClusterSpec
+	Status ClusterStatus
+}
+
+type ClusterSpec struct {
+	// Nominal version of Kubernetes control-plane to run.
+	KubernetesVersion KubernetesVersionInfo
+
+	// Basic API configuration
+	APIConfig APIServerConfig
+
+	// Cluster network configuration
+	ClusterNetwork ClusterNetworkingConfig
+
+	// Provider-specific serialized configuration to use during
+	// cluster creation. It is recommended that providers maintain
+	// their own versioned API types that should be
+	// serialized/deserialized from this field.
+	ProviderConfig string
+}
+
+type APIServerConfig struct {
+	// The address for the API server to advertise.
+	AdvertiseAddress string
+
+	// The port on which the API server binds.
+	Port uint32
+
+	// Extra Subject Alternative Names for the API server's serving cert.
+	ExtraSANs []string
+}
+
+type ClusterNetworkingConfig struct {
+	// The subnet from which service VIPs are allocated.
+	ServiceSubnet string
+
+	// The subnet from which POD networks are allocated.
+	PodSubnet string
+
+	// Domain name for services.
+	DNSDomain string
+}
+
+type KubernetesVersionInfo struct {
+	// Semantic version of Kubernetes to run.
+	Version string
+}
+
+type ClusterStatus struct {
+	// APIEndpoint represents the endpoint to communicate with the IP.
+	APIEndpoint APIEndpoint
+
+	// A simple boolean to indicate whether the control plane was
```

Comment:
What happens when you update the kubernetes version of the control plane but it has not yet reconciled that to a reality? Does this bool flip back to false?

Reply:
That's a good question. I wanted a simple way of determining when it was done provisioning, but this is probably too simple. Health of the cluster should probably be determined through the API endpoint.

Any other ideas of how to do this?

Reply:
You could have (A) a bool that flips back and forth depending on how reconciled the object is (like @pipejakob did in the machine representation) or (B) perhaps status can somehow have the version of kubernetes. Empty for not set up yet and the current version for what is currently up. (C) Taking (B) to the extreme would be to place a cluster spec in the cluster status. When the two match, everything is fully reconciled. When they do not match, you can see exactly what is the difference between goal and reality.

Reply:
I definitely think the current reconciled version should be stored in the status. Otherwise, controllers built against will receive an update to the `Cluster` resource, that has the k8s version updated to something new, and the `ready` bool still true (whilst the controller responsible for flipping the bool is still reconciling)